Documentation of the logging system (#1129)
dlr-gtlab/gtlab-core: docs/modules/logging.rst

    // WARN  [11-12-42] The required tolerance could not be achieved!
    gtError() << "Missing object 'Engine'";
    // ERROR [11-12-42] Missing object 'Engine'
    gtFatal() << "The CPU is malfunctioning. Aborting.";
    // FATAL [11-12-42] The CPU is malfunctioning. Aborting.

Each log message is accompanied by a timestamp and the moduleID in GTlab's output widget.

By employing these logging functions appropriately, developers can ensure comprehensive logging coverage within their modules, aiding in debugging and system monitoring.

Debugging
^^^^^^^^^

In GTlab, developers have the option to log additional information for debugging purposes,
aiding in the identification and resolution of problems within the framework.
By default, GTlab does not log debugging messages.
However, developers can enable the display of debug messages by adjusting the settings within GTlab.

For debugging purposes, GTlab provides the two functions ``gtDebug`` and ``gtTrace``, which can be used as follows:

.. code-block:: cpp

    gtDebug() << "Iteration count = " << count;
    // DEBUG [11-12-42] Iteration count = 101
    gtTrace() << "Begin MyTask::run";

>     // TRACE [11-12-42] My fancy trace message

Reviewer: ```suggestion
    // TRACE [11-12-42] Begin MyTask::run
```

Logging
=======

In GTlab, a comprehensive logging framework is provided to assist module developers in effectively
capturing and managing log messages from various components of their applications.

This section explores the logging capabilities offered by GTlab, including configuration options,
logging levels, custom logging IDs, support for custom types, and integration with external libraries
such as Qt and the C++ Standard Template Library (STL).

Basic Logging
-------------

Preparation
^^^^^^^^^^^

When using GTlab's logging library, module developer need to define the preprocessor variable

> ``GT_MODULE_ID`` accordingly. Settings this variable up is part of the module building instructions

Reviewer 2: ```suggestion
``GT_MODULE_ID`` accordingly. Settings this variable is part of the module building instructions
```

  This level of logging can be extremely verbose and is typically reserved for troubleshooting complex issues.

Logging Functions
^^^^^^^^^^^^^^^^^

GTlab provides four primary functions that module developers should utilize
to log messages effectively. These functions, namely ``gtFatal``, ``gtError``, ``gtWarning``,
and ``gtInfo``, are used as follows:

.. code-block:: cpp

    gtInfo() << "The residuum of the solver is r = " << residuum;
    // INFO  [11-12-42] The residuum of the solver is r = 0.1
    gtWarning() << "The required tolerance could not be achieved!";
    // WARN  [11-12-42] The required tolerance could not be achieved!
    gtError() << "Missing object 'Engine'";
    // ERROR [11-12-42] Missing object 'Engine'
    gtFatal() << "The CPU is malfunctioning. Aborting.";
    // FATAL [11-12-42] The CPU is malfunctioning. Aborting.

Each log message is accompanied by a timestamp and the moduleID in GTlab's output widget.

By employing these logging functions appropriately, developers can ensure comprehensive logging coverage within their modules, aiding in debugging and system monitoring.

**Logging messages only once**

> To make a message appear only once, use `gtLogOnce` together with the log level

Reviewer 2: It's good that the functionality is included. Can we explain what it does in more detail here? Perhaps a slightly larger example?

    but may result in increased log output and resource usage.
    Hence, use these statements sparingly in productive code.
    Use ``gtDebug``, to print values of variables to help understanding, what the program is doing.
    Use ``gtTrace`` only, to add information on the control flow of your program.
    If unsure, use ``gtDebug``.

**Line Numbers**

The logging system in GTlab offers the option to include the file and line number of the logging statement in the output.
This can be enabled by defining the preprocessor macro ``GT_LOG_LINE_NUMBERS``.

.. code-block:: cpp

    gtDebug() << "My fancy message";
    // DEBUG [11-12-42] my_file.cpp@123: My fancy message

Note, that line number should only be enabled in debug mode!

Advanced Usage
--------------

Custom Logging IDs
^^^^^^^^^^^^^^^^^^

When using the logging functions in GTlab, the module ID is employed as a logging ID.
To enhance the ability to filter specific messages from subcomponents within a module,
custom logging IDs can be specified using the logging functions ``gtFatalId``, ``gtErrorId``,
``gtWarningId``, ``gtInfoId``, ``gtDebugId``, and ``gtTraceId``.

For example

.. code-block:: cpp

>     gtWarningId("SolverComponent") << "The solver has not converged";

Reviewer 2: ```suggestion
    gtWarningId("SolverComponent") << "The solver has not converged";
    // WARN [10:38:46] [SolverComponent] The solver has not converged
```

.. note::

    If you are using CMake as your build system, ``GT_LOG_USE_QT_BINDINGS`` is already defined
    by the GTlab::Core target. You don't have to do anything.

Extended STL Support
^^^^^^^^^^^^^^^^^^^^

By default, the logging library does not provide logging operators for most STL classes (such as vector, list, tuple, etc.)
to maintain a lean logging interface. However, similar to enabling Qt Support, developers can enable extended
STL support by setting a global define ``GT_LOG_USE_EXTENDED_STL_BINDINGS``.

Alternatively, developers can include specific header files to enable extended STL support for individual STL classes. For example:

- Including `gt_logging/stl_bindings.h` enables extended STL support for all available STL classes.
- Including specific header files like `gt_logging/vector.h` enables extended STL support specifically for `std::vector`.

.. note::

    If Qt support is enabled either globally or by including Qt bindings, all available STL bindings will be included automatically

Logging of Custom Types
^^^^^^^^^^^^^^^^^^^^^^^

> To implement logging for a custom type, a custom ``operator<<`` must be defined for this type.
> It must be defined in the global namespace:

Reviewer 2: ```suggestion
To implement logging for a custom type, a corresponding custom``operator<<`` must be defined. This operator must be located in the global namespace:
```
